TripalCultivate/TripalCultivate-Phenotypes
# G4.114 Finalize processValidationMessages() to generate user-friendly messages

**Issue #114**

**PR #123 has been merged into this PR, please see that PR for context.**
This PR _solely_ covers changes done to: 
  - [TripalCultivatePhenotypesTraitsImporter.php](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/127/files#diff-67c6ba893a67e6b0852eebeafa9c18ea19db7d7c7fa68a22530f746f526e89b1)
  - [TraitImporterFormValidateTest.php](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/127/files#diff-48c14597475a5b9b2e678d0744a866aa9a8fd8f8697843d7996dc93704e349d2)

## Motivation

This PR is meant to tackle changes needed to processValidationMessages() in the Traits Importer. Once the feedback validation engine has been upgraded to expect a render array to display validation errors to the user, then processValidationMessages() is updated to create those render arrays for each validator that fails.

## What does this PR do?
1. For each validator that is instantiated by the Traits Importer, create a new method that chooses the appropriate render array type/markup, takes the contents of `$failures[$validator_name]` and plugs them into the appropriate places of the render array, and finally returns the finished render array.
   - [x] `processGenusExistsFailures()`
   - [x] `processValidDataFileFailures()`
   - [x] `processValidHeadersFailures()`
   - [x] `processValidDelimitedFileFailures()`
   - [x] `processEmptyCellFailures()`
   - [x] `processValueInListFailures()`
   - [x] `processDuplicateTraitsFailures()`

2. Update `processValidationMessages()` to no longer loop through all of the failed validators, and instead checks each validator if a failure occurred and calls the correct method to build the render array.

- [x] Update `TraitImporterFormValidateTest` to check for expected elements in the render arrays (ie. contents of 'details') for each scenario.

- [x] Creates new tests in `TraitImporterProcessValidationTest` to trigger every case from the following methods and render the output to compare with expectations. Each test will have its own data provider.
   - [x] `testProcessGenusExistsFailures()`
   - [x] `testProcessValidHeadersFailures()`
   - [x] `testProcessDuplicateTraitsFailures()`

## Testing

### Automated Testing
No new automated tests have been added yet, though some will need to be added to maintain 100% coverage of the new process methods.

### Manual Testing
Thanks to PR #123, we can now test this through the UI!
1. Create a new docker container from this branch.
2. Open the site in a browser and log in.
3. Configure the site by:
   - Creating a new organism (Tripal > Content > Add Tripal Content)
   - Configuring the module (Tripal > Extensions > Tripal Cultivate Phenotypes Configuration > Ontology Terms)
4. Go to Tripal > Data Loaders > Tripal Cultivate: Phenotypic Trait Importer
5. Upload test files (see: `trpcultivate_phenotypes/tests/src/Fixtures/TraitImporterFiles` for the test files used by `traitImporterFormValidateTest`) and see if the validation output is intuitive for each test case. Note that validators run once (ie. GenusExists, ValidDataFile, ValidHeaders) will show an itemized list for failed validation, whereas row-level validators (ie. ValidDelimitedFile, EmptyCell, ValueInList, DuplicateTraits) will display a table with a row per failed validation of a line in the input file. Feel free to get creative and modify the test files to try different scenarios.
